pkg/integrations/register.go

```diff
@@ -0,0 +1,206 @@
+// will represent the configuration that controls the specific integration.
+// Registered IntegrationConfigs may be loaded using UnmarshalYAML or manually
+// constructed.
+//
+// RegisterIntegration panics if cfg is not a pointer.
+func RegisterIntegration(cfg IntegrationConfig) {
+	if reflect.TypeOf(cfg).Kind() != reflect.Ptr {
+		panic(fmt.Sprintf("RegisterIntegration must be given a pointer, got %T", cfg))
+	}
+	registeredIntegrations = append(registeredIntegrations, cfg)
+}
+
+// IntegrationConfig provides the configuration and constructor for an
+// integration.
+type IntegrationConfig interface {
+	// Name returns the name of the integration and the key that will be used to
+	// pull the configuration from the Agent config YAML.
+	Name() string
+
+	// IsEnabled returns whether this integration should run.
+	IsEnabled() bool
+
+	// NewIntegration returns an integration for the given with the given logger.
+	NewIntegration(l log.Logger) (common.Integration, error)
+}
+
+// IntegrationConfigs is a list of integrations.
+type IntegrationConfigs []IntegrationConfig
+
+func (c *IntegrationConfigs) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	return c.unmarshalWithIntegrations(registeredIntegrations, unmarshal)
+}
+
+func (c *IntegrationConfigs) unmarshalWithIntegrations(integrations []IntegrationConfig, unmarshal func(interface{}) error) error {
+	// Create a dynamic struct type full of our registered integrations and
+	// unmarshal to it.
+	var fields []reflect.StructField
+	for _, cfg := range integrations {
+		fields = append(fields, reflect.StructField{
+			Name: "Config_" + cfg.Name(),
+			Tag:  reflect.StructTag(fmt.Sprintf(`yaml:"%s"`, cfg.Name())),
+			Type: reflect.TypeOf(cfg),
+		})
+	}
+
+	var (
+		structType = reflect.StructOf(fields)
+		structVal  = reflect.New(structType)
+	)
+	if err := unmarshal(structVal.Interface()); err != nil {
+		return err
+	}
+
+	// Go over all non-nil fields in structVal and append them to c.
+	structVal = structVal.Elem()
+	for i := 0; i < structVal.NumField(); i++ {
+		if structVal.Field(i).IsNil() {
+			continue
+		}
+
+		val := structVal.Field(i).Interface().(IntegrationConfig)
+		*c = append(*c, val)
+	}
+
+	return nil
+}
+
+// UnmarshalYAML helps implement yaml.Unmarshaller for structs that have an
+// IntegrationConfigs field that should be inlined in the YAML string.
+func UnmarshalYAML(out interface{}, unmarshal func(interface{}) error) error {
+	return unmarshalIntegrationsWithList(registeredIntegrations, out, unmarshal)
+}
+
+// unmarshalIntegrationsWithList unmarshals to a subtype of out that has a
+// field added for every integration in integrations. Code adapted from
+// Prometheus:
+//
+//   https://github.com/prometheus/prometheus/blob/511511324adfc4f4178f064cc104c2deac3335de/discovery/registry.go#L111
+func unmarshalIntegrationsWithList(integrations []IntegrationConfig, out interface{}, unmarshal func(interface{}) error) error {
```

Comment:
This function makes me sad but there's really no other way to handle this. We need strict unmarshaling to still work, so we have to do what prometheus does: create a merged struct of the target + all of the possible integration fields, copy defaults to the merged struct, unmarshal to the merged struct, copy original fields back to the output, and then fill in the integrations slice. 

I'll try to document this more because my original look at Prometheus' code had me really confused about what was going on and why.